### ISSUE TITLE
Windows Named Pipe Monitor: Ability to filter on PID or file ID 

### ISSUE THREAD

**Issue by mikeparker** [vovkos/ioninja-issues#7]
I am trying to examine traffic over a named pipe between multiple processes, however the requests are interleaved in the log file and its extremely difficult to consolidate a single call to see whats going on, see the image for an example:
![image](https://user-images.githubusercontent.com/1557887/63364920-6cc24280-c36e-11e9-9751-41be1d056f0f.png)

Can we get a way to consolidate requests from a particular PID or File ID?

**Comment by vovkos** [comment]
Yes, solving this problem (and similar issues in TCP/UDP Flow Monitor, TCP Listener/Proxy, etc) is on our TODO list. One approach we consider is this: a user needs to be able to click a record inside the log and *extract* the corresponding stream into a separate session/log -- we will call this feature "session forking". Another approach is to implement post-capture filtering -- now we have filename filter which is applied at capture-time; there should be a second level of filtering which would allow to narrow down the log *after* everything was captured.

Until it's done, here's what you can do right now:

(1) Implement a custom filter layer (use ``scripts/plugins/TxRxFilter`` as an example). To filter pipe log based on PID/FileID, you would need to specially process these record codes: ``PipeMonLogRecordCode.ServerFileOpened``, ``PipeMonLogRecordCode.ClientFileOpened``, ``PipeMonLogRecordCode.FileIdChanged`` (defined in ``scripts/plugins/PipeMon/PipeMonLog.jnc``). These records contain PID/FileID identifiers. Check those against user-specified ID values, and if it's a mismatch, ignore the upcoming records until the you see a matching PID/FileID pair.

(2) Instead of creating a plugin for IO Ninja, you can post-process an ``*.njlog`` file using another language you are comfortable with (e.g. Python). The definitions for log file structs can be found in: ``scripts/api/log_RecordFile.jnc`` and ``scripts/api/log_RecordCode.jnc``; an example of iterating through all the records in an ``*.njlog`` file can be found in ``scripts/plugins/ReplayLog``. The logic of post processing is the same as in (1) -- you look for ``PipeMonLogRecordCode.ServerFileOpened``, ``PipeMonLogRecordCode.ClientFileOpened``, ``PipeMonLogRecordCode.FileIdChanged`` and throw away records with mismatching PIDs/FileIDs

**Comment by vovkos** [comment]
I just thought of a simple hack which might be just enough in your case.

Append this to the end of ``scripts/plugins/PipeMon/PipeMonLog.njplg``: 
```
[filter]

id          = file-id-changes
name        = "File ID changes"
description = "Show/hide file ID changes"

01d48717897030ae  # FileIdChanged
```

Restart IO Ninja, open a PipeMon session and then add the TxRxFilter layer. Then go to the Settings and uncheck``File ID Changes`` and *one of* ``Show TX``/``Show RX``. Click ``Apply and Rebuild Log`` and you will have a much cleaner log.

**Comment by mikeparker** [comment]
I can't see `File ID Changes` in the settings. Here's my file

```
#...............................................................................
#
#  Pipe Monitor plugin
#

[plugin]

name        = "Pipe Monitor"
description = "Monitor named pipe communications"
id          = pipemon
group       = fs
monitor     = true
log-guid    = 8D4114D9-5020-4E0D-8D6A-98C8C29E76F0  # pipe monitor
log-guid    = 323abaeb-7859-48bc-8e24-f8c082b04ff7  # device monitor
icon        = images/pipemon.png
small-icon  = images/pipemon-16x16.png

session-class = PipeMonSession

[scripts]

PipeMonSession.jnc
PipeMonitor.jnc

[filter]

id          = file-id-changes
name        = "File ID changes"
description = "Show/hide file ID changes"

01d48717897030ae  # FileIdChanged
```

![image](https://user-images.githubusercontent.com/1557887/63432935-0f83cb00-c41a-11e9-93e4-b6994ab9c1f3.png)

**Comment by vovkos** [comment]
You need to modify ``PipeMonLog.njplg``, not ``PipeMon.njplg``

**Comment by mikeparker** [comment]
Ok, think this is working, although I'm getting garbled messages like its 2 files combined, is this correct?

![image](https://user-images.githubusercontent.com/1557887/63434548-3abbe980-c41d-11e9-937c-c95c1b373657.png)

**Comment by vovkos** [comment]
Can you save and attach the original log file (``*.njlog``)? It looks like there is more than one RX stream in it.

**Comment by mikeparker** [comment]
I dont know if this screws up the formatting. Let me know if i need to upload the file somewhere

https://pastebin.com/N63PWm8v

**Comment by mikeparker** [comment]
[test.log](https://github.com/vovkos/ioninja-issues/files/3525507/test.log)

**Comment by mikeparker** [comment]
I'm wondering if theres a further bug whereby either messages get duplicated or the tx/rx (or client/server) get mixed up somehow? I'm seeing logs like this

```
PID #13656 file #2: 
48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0a  HTTP/1.1 200 OK.
PID #13656 file #10: 
48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0a  HTTP/1.1 200 OK.
```
First one is Tx second one is Rx. 

The majority of requests are in blue, and responses are in green. But every now and then theres a request in green and a response in blue.

**Comment by vovkos** [comment]
I checked the log; there is a mix of multiple file streams, so of course, simple filtering by TX/RX will not yield the desired result (I suggested this hack because in the very first screenshot I only saw two files, a client and a server).

Re duplication of packets -- normally, every packet can be seen twice: first, when it's written, and the second time when it's read. This applies to both requests and responses. It's not a bug, that's just how NPFS (named pipe file system) works -- first, data is written to one file, then the same data is read from another file; both files satisfy the filename filter.

So... I would recommend going back to the solutions suggested in my first reply (e.g. write a program in C or Python or whatever works for you -- to process the ``*.njlog`` file and extract per-file data streams). We will definitely implement this secondary PID/FID filter in the Pipe Monitor plugin soon, but it might be faster for you to just write a filtering program yourself.

**Comment by mikeparker** [comment]
> I checked the log; there is a mix of multiple file streams, so of course, simple filtering by TX/RX will not yield the desired result (I suggested this hack because in the very first screenshot I only saw two files, a client and a server).

Ok no problem, thanks for your time.

> Re duplication of packets -- normally, every packet can be seen twice: first, when it's written, and the second time when it's read. This applies to both requests and responses. It's not a bug, that's just how NPFS (named pipe file system) works -- first, data is written to one file, then the same data is read from another file; both files satisfy the filename filter.

So would you expect to see every single character going down the named pipe appear twice? Because that's not what is appearing on the log I am looking at. Is the log viewer hiding some reads somehow? I suspect it's visible in the log I posted earlier - the majority of requests and responses only appear once. I know the request data has been read because the response is related to the data from the request.

Examples
![image](https://user-images.githubusercontent.com/1557887/63439894-d00fab80-c426-11e9-9bc1-0bdeef0aaed4.png)

![image](https://user-images.githubusercontent.com/1557887/63439911-d867e680-c426-11e9-8312-03148770169b.png)

(I know the screenshots don't prove those values don't appear anywhere else, but they don't)

**Comment by vovkos** [comment]
You will not see the reads or writes for files which were opened *before* you started the capture. You see, your log starts with a *client open*:

![pipemon-log](https://user-images.githubusercontent.com/1272029/63444086-9cd91680-c43e-11e9-9ac2-ca0915ef1e19.png)

it means, a server-side file has already been open before. Reads and writes to this server file will never be seen in this log.

That being said, I totally understand confusion a real-life pipe log brings. We definitely need to improve the usablity of our pipe monitor by allowing to conveniently extract per-file streams (to a separate view, or at least as a secondary filter).

**Comment by mikeparker** [comment]
Right ok, yep I see now if i start the log before i start the server i get both sides. Thanks. I'm going off topic so I will raise another issue or two about a few other bits that might be nice.

**Comment by mikeparker** [comment]
Should the file ID be the same number on Tx and Rx then? It looks like they are different numbers here which is a bit confusing. 17 -> 22

![image](https://user-images.githubusercontent.com/1557887/63446649-7c0ac400-c432-11e9-9760-3c9843a7220c.png)

Also why is only one side of the file 'connection accepted'?
![image](https://user-images.githubusercontent.com/1557887/63446894-ecb1e080-c432-11e9-904f-170706b17760.png)

**Comment by vovkos** [comment]
> Should the file ID be the same number on Tx and Rx then? It looks like they are different numbers here which is a bit confusing. 17 -> 22

Well, these are two different ``FILE_OBJECT``-s (one is from the server, one is from the client). I can't think of any reliable way to *deduce a connection* between these two (both client and server can open multiple ``FILE_OBJECT``-s with *the same name*, and only the ``npfs.sys`` driver knows how they are really interconnected.

> Also why is only one side of the file 'connection accepted'?

Because only the server side does this extra "accepting" step. The client side just calls ``CreateFile`` to connect, while the server side first opens a named pipe with ``CreateNamedPipe`` and then calls ``ConnectNamedPipe`` to accept a client. The *completion* of the ``ConnectNamedPipe`` request is manifested by this "Connection accepted" message.

**Comment by vovkos** [comment]
We are about to release a major update for IO Ninja; among many other things, it has a filter for named pipe names/file ids/process ids which can be applied *after* the capture. Take a look:

![pipemon-new](https://user-images.githubusercontent.com/1272029/70316859-88872100-1857-11ea-9080-f4f034fe095f.png)

The link to the internal build:
https://tibbo.com/downloads/archive/ioninja/.internal/ioninja-3.11.10-windows-amd64.7z

**Comment by vovkos** [comment]
The feature is implemented in the just-released ioninja-3.12.0
https://ioninja.com/downloads.html
